File: advertising_sdk/src/main/java/com/adadapted/android/sdk/ui/messaging/AdContentPublisher.java
package com.adadapted.android.sdk.ui.messaging;

import android.os.Handler;
import android.os.Looper;

import com.adadapted.android.sdk.ui.model.AdContent;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class AdContentPublisher {
    @SuppressWarnings("unused")
    private static final String TAG = AdContentPublisher.class.getName();

    private static AdContentPublisher sAdContentPublisher;

    public static synchronized AdContentPublisher getInstance() {
        if(sAdContentPublisher == null) {
            sAdContentPublisher = new AdContentPublisher();
        }

        return sAdContentPublisher;
    }

    private final Set<AdContentListener> listeners;
    private final Lock lock = new ReentrantLock();

    private AdContentPublisher() {
        listeners = new HashSet<>();
    }

    public void addListener(final AdContentListener listener) {
        lock.lock();
        try {
            if (listener != null) {
                listeners.add(listener);
            }
        }
        finally {
            lock.unlock();
        }
    }

    public void removeListener(final AdContentListener listener) {
        lock.lock();
        try {
            if(listener != null) {
                listeners.remove(listener);
            }
        }
        finally {
            lock.unlock();
        }
    }

    public void publishContent(final String zoneId,
                               final AdContent content) {
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                lock.lock();
                try {
                    for(final AdContentListener listener : listeners) {
                        listener.onContentAvailable(zoneId, content);
                    }
                }
                finally {
                    lock.unlock();
                }
            }
        });
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/ui/messaging/AdContentPublisher.kt
package com.adadapted.android.sdk.ui.messaging

import android.os.Handler
import android.os.Looper
import com.adadapted.android.sdk.core.ad.AdContent
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

class AdContentPublisher private constructor() {
    private val listeners: MutableSet<AdContentListener> = HashSet()
    private val lock: Lock = ReentrantLock()

    fun addListener(listener: AdContentListener) {
        lock.lock()
        try {
            listeners.add(listener)
        } finally {
            lock.unlock()
        }
    }

    fun removeListener(listener: AdContentListener) {
        lock.lock()
        try {
            listeners.remove(listener)
        } finally {
            lock.unlock()
        }
    }

    fun publishContent(zoneId: String, content: AdContent) {
        if (content.hasNoItems()) {
            return
        }
        Handler(Looper.getMainLooper()).post {
            lock.lock()
            try {
                for (listener in listeners) {
                    listener.onContentAvailable(zoneId, content)
                }
            } finally {
                lock.unlock()
            }
        }
    }

    companion object {
        private val LOGTAG = AdContentPublisher::class.java.name
        private lateinit var instance: AdContentPublisher

        fun getInstance(): AdContentPublisher {
            if (!this::instance.isInitialized) {
                createInstance()
            }
            return instance
        }

        fun createInstance() {
            instance = AdContentPublisher()
        }
    }
}